rclcpp/CMakeLists.txt

```diff
@@ -109,6 +109,7 @@ ament_target_dependencies(${PROJECT_NAME}
   "rosidl_typesupport_cpp"
   "rosidl_generator_cpp")
 
+configure_rcl(${PROJECT_NAME})
```

Comment:
Needs to be rclcpp?

Reply:
I am trying to reuse the **configure_rcl** macro in ros2/rcl. This is in a way similar to how the [configure_rmw_library](https://github.com/ros2/rmw/blob/master/rmw/cmake/configure_rmw_library.cmake) macro is being reused in [rmw_fastrtps](https://github.com/ros2/rmw_fastrtps/blob/328771c18382bd053d4a6eb34992914df40fe3c4/rmw_fastrtps_cpp/CMakeLists.txt#L99) and other packages. The single macro takes care of both C and C++ code

Reply:
This is not possible because they are not part of the same shared library. See doc here: https://gcc.gnu.org/wiki/Visibility

Reply:
A possible alternative may be to use the CMake standard macro [`generate_export_header`](https://cmake.org/cmake/help/v3.5/module/GenerateExportHeader.html) and the [`DEFINE_SYMBOL`](https://cmake.org/cmake/help/v3.5/prop_tgt/DEFINE_SYMBOL.html) target property.

See https://github.com/ros/console_bridge/pull/43 for a use in a ROS-related repo, and https://bitbucket.org/osrf/gazebo/issues/2262/symbol-visibility-issues-in-windows for a related issue in Gazebo.

Reply:
Quick note: This is the way in which we are sharing the "configure_rcl" macro from rcl to rclcpp: https://github.com/ros2/rmw/blob/master/rmw/CMakeLists.txt#L62-L65
I am not sure if sharing just the macro is allowing librcl_cpp.so to access librcl.so private symbols.

Reply:
The critical part is in https://github.com/ros2/rcl/pull/391/files#diff-19c13a4e66284bae327c94111b226946R71, in which you are defining the `RCL_BUILDING_DLL` macro also for compilation units that are not part of the `rcl`. The effect will be that when including https://github.com/ros2/rcl/blob/aa4ac7d06cc5e59d693408260868edb004154b69/rcl/include/rcl/visibility_control.h as part of a compilation unit **outside of rcl**, `RCL_PUBLIC` (in Windows) will have the value `__declspec(dllexport)` instead of the correct `__declspec(dllimport)`. This is exactly the problem described in https://bitbucket.org/osrf/gazebo/issues/2262/symbol-visibility-issues-in-windows .

Reply:
> A possible alternative may be to use the CMake standard macro [`generate_export_header`](https://cmake.org/cmake/help/v3.5/module/GenerateExportHeader.html) and the [`DEFINE_SYMBOL`](https://cmake.org/cmake/help/v3.5/prop_tgt/DEFINE_SYMBOL.html) target property.

I personally don't like the generated code approach (I prefer a single, checked-in boil-plate file) and a custom define.

> I am trying to reuse the **configure_rcl** macro in ros2/rcl. This is in a way similar to how the [configure_rmw_library](https://github.com/ros2/rmw/blob/master/rmw/cmake/configure_rmw_library.cmake) macro is being reused in [rmw_fastrtps](https://github.com/ros2/rmw_fastrtps/blob/328771c18382bd053d4a6eb34992914df40fe3c4/rmw_fastrtps_cpp/CMakeLists.txt#L99) and other packages. The single macro takes care of both C and C++ code

As the others said, you can do this, but only for the visibility compiler options, not the visibility compiler definition.

Reply:
Thanks for the inputs @thomas-moulard , @traversaro and @wjwwood. I have removed the target definition specific code in [configure_rcl](https://github.com/ros2/rcl/pull/391/files#diff-19c13a4e66284bae327c94111b226946R31) in https://github.com/ros2/rcl/pull/391 . This is because each rcl or rclcpp has its own target definition specified in the CMakeList. For eg:
https://github.com/ros2/rclcpp/blob/master/rclcpp/CMakeLists.txt#L112-L115
https://github.com/ros2/rclcpp/blob/master/rclcpp_action/CMakeLists.txt#L38-L41

Is this a reasonable solution? Do you think we should have a separate configure_rclcpp for rclcpp packages?

Reply:
> I personally don't like the generated code approach (I prefer a single, checked-in boil-plate file) and a custom define.

This is probably a bit OT in this PR, but for the sake of clarity: the generated header file (both CMake's and the example linked by @thomas-moulard ) is useful for supporting the case in which a library is compiled as a static library under Windows. Without that, in any downstream compilation unit `RCLCPP_PUBLIC` is always defined as `__declspec(dllimport)` and this annotation will generate a Visual Studio error LNK2019 if the RCLCPP library is a actually compiled as static. A possible alternative **without** having a generated file is to have an additional logic that defines `RCLCPP_PUBLIC` as an empty macro if `RCLCPP_STATIC` is defined, but then it is necessary some kind of build system logic to make sure that `RCLCPP_STATIC` is always defined (even downstream) if RCLCPP is build as static, for example a conditional `target_compile_definitions(rclcpp PUBLIC RCLCPP_STATIC)` called only if the library is built is static, or similar.

See https://github.com/ros/console_bridge/pull/40#issuecomment-269460721 for a similar comment.